lib/ansible/template/safe_eval.py

```diff
@@ -19,141 +19,164 @@
 __metaclass__ = type
 
 import ast
-import sys
+import copy
+
+from jinja2 import StrictUndefined, UndefinedError
 
-from ansible import constants as C
 from ansible.module_utils.common.text.converters import container_to_text, to_native
-from ansible.module_utils.six import string_types, PY2
-from ansible.module_utils.six.moves import builtins
-from ansible.plugins.loader import filter_loader, test_loader
+from ansible.module_utils.six import PY2, string_types
+from ansible.utils.display import Display
+
+display = Display()
+
+
+def ansible_undefined(*args, **kwargs):
+    """Short circuit to raise undefined when evaling ``AnsibleUndefined``"""
+    bool(StrictUndefined(*args, **kwargs))
+
+
+# Define globals that we understand and can convert to a Python type
+# or are used to limit functionality
+_OUR_GLOBALS = {
+    '__builtins__': {},  # avoid global builtins as per eval docs
+
+    # JSON: These are included to automatically convert JSON objects to python
+    'false': False,
+    'null': None,
+    'true': True,
+
+    # Callables
+    'AnsibleUndefined': ansible_undefined,
+
+    # Example for calling staticmethod/classmethod
+    # 'Foo': Foo,
```

Comment:
If we don't want `Foo` to also be callable, we'll need to maintain `_OUR_GLOBALS` and `_CALL_ENABLED` separately.